### PR TITLE
iter6/7 - Maven Support with File Reading and Writing

### PR DESCRIPTION
**Pull request by jinocencio-solvd**

### CODE REVIEW

**Comment by jinocencio-solvd** [comment]
### Task - Maven Support

Requirements:

- Move all projects to Maven.
  - Project initially created using Intellij Maven Java Project starter
  - Alternatively, we could have used `mvn archetype:generate` to initialize the project
    - ```
      mvn archetype:generate 
      -DgroupId=org.example 
      -DartifactId=PharmacyApp
      -DarchetypeArtifactId=maven-archetype-quickstart 
      -DarchetypeVersion=1.4 
      -DinteractiveMode=false
      ```

- Build jar file and deploy to the local repository.
  - ```
    mvn deploy:deploy-file 
    -Dfile=target/PharmacyApp-1.0-SNAPSHOT.jar
    -DgroupId=org.example
    -DartifactId=PharmacyApp
    -Dversion=1.0-SNAPSHOT
    -Dpackaging=jar
    -Durl=file://~/.m2/repository/
    ```
  - In order to build a jar file for our app that includes the `log4j2` dependency, we have to add the Maven plugins `maven-assembly-plugin` and `maven-jar-plugin` into the `pom.xml` file
- Add and use 2 Plugins.
    - `maven-assembly-plugin` - used to create JAR files with all dependencies included.
    - `maven-jar-plugin` - creates executable JAR file with manifest configuration and location of the application entry point.
    - `maven-surefire-plugin` - used to run tests during the build process.
- Run mvn for different phases from the Maven lifecycle. Check the result.
  - clean - removes generated `target` folder from previous build.
  - test - runs all test classes and methods in `src/test`.
  - package - packages the application into a JAR file.
  - site - generates project's site documentation in the `target/site` directory.

File: src/main/java/misc/DataProvider.java [hunk 1]
```diff
@@ -101,4 +102,12 @@ public static Pharmacy predefinedPharmacy() {
         return new Pharmacy("PharmacyRx", predefinedAddresses()[0], "(123)-342-4344",
             "pharm@pharmaEmail.com");
     }
+
+    public static Cart predefinedCartItems() {
```
Comment (ankochergin) [inline src/main/java/misc/DataProvider.java]:
```suggestion
    public static Cart predefinedCartWithItems() {

```

CartItems sounds like it returns a List of Items, not a Cart

File: src/main/resources/log4j2.xml [hunk 2]
```diff
@@ -9,14 +9,14 @@
...
-    </File>
+<!--    <File name="file" fileName="PharmacyAppLogs.log" append="false">-->
+<!--      <PatternLayout pattern="%d{HH:mm:ss.SSS} %-5level %logger{36}.%M - %msg%n" disableAnsi="false"/>-->
+<!--    </File>-->
   </Appenders>
   <Loggers>
     <Root level="trace" additivity="false">
       <AppenderRef ref="console"/>
-      <AppenderRef ref="file"/>
+<!--      <AppenderRef ref="file"/>-->
```
Comment (ankochergin) [inline src/main/resources/log4j2.xml]: uncomment these lines, just don't upload your log file

File: src/main/java/misc/DataProvider.java (hunk 1 above)
Comment (jinocencio-solvd) [reply src/main/java/misc/DataProvider.java]: I understand your point. I changed and renamed the method to return carts array to be more consistent with my other DataProvider methods.

**Comment by jinocencio-solvd** [comment]
## Iter-7

### Task - File Reader

Requirements:
- Read text from the file and calculate the numbers of the unique words. Write the result to the file. The main requirement is: using StringUtils and FileUtils to implement it with minimum lines of code.
  - The `StringUtils.split` method was used to convert the contents of the text file into an array of strings which can be processed to count the number of unique words. This method is similar to using the `String.split` method, but is provided by the Apache Commons Lang library.
  - `FileUtils.readFileToString` method was used to read the contents of the text file and convert it to a string. This method simplifies the process of reading the file by avoiding the need to manually loop through the file and read individual lines using the `BufferedReader` and `FileReader` approach.
  - To write the output to a file, the `FileUtils.writeStringToFile` method was used. This method allows us to write a customized message to the file, with the option to append the message to an existing file. The append parameter is set to true to append the message to the existing file, and avoids the need to initialize a `FileWriter` with the `append` parameter set to `true`, manually writing the message to the file `FileWriter.write`, and then closing the writer `FileWriter.close`.

File: src/main/java/Main.java [hunk 3]
```diff
@@ -72,7 +71,7 @@ private static void userCreatePatient(Pharmacy pharmacy) {
         }
     }
 
-    public static void main(String[] args) throws DuplicatePersonException {
+    private static void pharmacyOperations() throws DuplicatePersonException {
```
Comment (ankochergin) [inline src/main/java/Main.java]: you can leave this as main and for String and FileUtils task have some countWords static method

File: src/main/java/Main.java [hunk 4]
```diff
@@ -99,4 +100,9 @@ public static void main(String[] args) throws DuplicatePersonException {
             customerLine.getNextCustomer();
         }
     }
+
+    public static void main(String[] args) throws DuplicatePersonException {
```
Comment (ankochergin) [inline src/main/java/Main.java]: why it throws `DuplicatePersonException`?

Comment (jinocencio-solvd) [reply src/main/java/Main.java]: Thank you for catching that. This was a fragment left over from development. It is handled by parent class and all of its references have been removed from Main

File: src/main/java/Main.java (hunk 3 above)
Comment (jinocencio-solvd) [reply src/main/java/Main.java]: refactored the main method to include these changes

**Pull request merged**